File: implementations/python/deques.py
"""
Double-Ended Queues (Deques)

This module provides reference implementations of:
- CircularDeque: Fixed-capacity circular buffer deque
- LinkedDeque: Doubly-linked list deque
- sliding_window_max: Monotonic deque algorithm for window maxima
"""

from typing import Any, List, Optional
from collections import deque
import unittest


class CircularDeque:
    """Fixed-capacity circular array deque."""
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("Capacity must be positive")
        self._data = [None] * capacity
        self._front = 0
        self._size = 0
        self._capacity = capacity

    def empty(self) -> bool:
        return self._size == 0

    def full(self) -> bool:
        return self._size == self._capacity

    def size(self) -> int:
        return self._size

    def push_front(self, value: Any) -> None:
        if self.full():
            raise OverflowError("Deque overflow")
        self._front = (self._front - 1 + self._capacity) % self._capacity
        self._data[self._front] = value
        self._size += 1

    def push_back(self, value: Any) -> None:
        if self.full():
            raise OverflowError("Deque overflow")
        idx = (self._front + self._size) % self._capacity
        self._data[idx] = value
        self._size += 1

    def pop_front(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        val = self._data[self._front]
        self._data[self._front] = None
        self._front = (self._front + 1) % self._capacity
        self._size -= 1
        return val

    def pop_back(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        idx = (self._front + self._size - 1) % self._capacity
        val = self._data[idx]
        self._data[idx] = None
        self._size -= 1
        return val

    def front(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        return self._data[self._front]

    def back(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        idx = (self._front + self._size - 1) % self._capacity
        return self._data[idx]


class LinkedDeque:
    """Doubly linked-list deque."""
    class _Node:
        def __init__(self, value: Any):
            self.value = value
            self.prev: Optional['LinkedDeque._Node'] = None
            self.next: Optional['LinkedDeque._Node'] = None

    def __init__(self):
        self._head: Optional[LinkedDeque._Node] = None
        self._tail: Optional[LinkedDeque._Node] = None
        self._size = 0

    def empty(self) -> bool:
        return self._size == 0

    def size(self) -> int:
        return self._size

    def push_front(self, value: Any) -> None:
        node = self._Node(value)
        node.next = self._head
        if self._head:
            self._head.prev = node
        else:
            self._tail = node
        self._head = node
        self._size += 1

    def push_back(self, value: Any) -> None:
        node = self._Node(value)
        node.prev = self._tail
        if self._tail:
            self._tail.next = node
        else:
            self._head = node
        self._tail = node
        self._size += 1

    def pop_front(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        val = self._head.value
        self._head = self._head.next
        if self._head:
            self._head.prev = None
        else:
            self._tail = None
        self._size -= 1
        return val

    def pop_back(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        val = self._tail.value
        self._tail = self._tail.prev
        if self._tail:
            self._tail.next = None
        else:
            self._head = None
        self._size -= 1
        return val

    def front(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        return self._head.value

    def back(self) -> Any:
        if self.empty():
            raise IndexError("Deque underflow")
        return self._tail.value


def sliding_window_max(nums: List[int], k: int) -> List[int]:
    """Computes sliding window maximum using a monotonic deque in O(n) time."""
    dq = deque()
    result = []
    n = len(nums)
    if n == 0 or k <= 0 or k > n:
        return result

    for i, x in enumerate(nums):
        while dq and dq[0] <= i - k:
            dq.popleft()

        while dq and nums[dq[-1]] <= x:
            dq.pop()

        dq.append(i)

        if i >= k - 1:
            result.append(nums[dq[0]])

    return result


# ============================================================================
# Unit Tests
# ============================================================================

class TestDeques(unittest.TestCase):
    def test_circular_deque(self):
        dq = CircularDeque(4)
        self.assertTrue(dq.empty())

        dq.push_back(10)
        dq.push_back(20)
        dq.push_front(5)
        dq.push_front(1)
        self.assertTrue(dq.full())
        self.assertEqual(dq.front(), 1)
        self.assertEqual(dq.back(), 20)

        self.assertEqual(dq.pop_front(), 1)
        self.assertEqual(dq.pop_back(), 20)
        self.assertEqual(dq.front(), 5)
        self.assertEqual(dq.back(), 10)

    def test_linked_deque(self):
        dq = LinkedDeque()
        dq.push_front("mid")
        dq.push_front("start")
        dq.push_back("end")
        self.assertEqual(dq.size(), 3)
        self.assertEqual(dq.front(), "start")
        self.assertEqual(dq.back(), "end")

        self.assertEqual(dq.pop_front(), "start")
        self.assertEqual(dq.pop_back(), "end")
        self.assertEqual(dq.pop_front(), "mid")
        self.assertTrue(dq.empty())

    def test_sliding_window_max(self):
        nums = [1, 3, -1, -3, 5, 3, 6, 7]
        k = 3
        res = sliding_window_max(nums, k)
        self.assertEqual(res, [3, 3, 5, 5, 6, 7])


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/deques.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <stdexcept>
#include <cassert>

namespace dsa {

/**
 * @brief Fixed-capacity circular buffer deque.
 * Supports push/pop at both front and back in O(1) time.
 */
template <typename T>
class CircularDeque {
public:
    explicit CircularDeque(size_t capacity)
        : data_(capacity), front_(0), size_(0) {
        if (capacity == 0) {
            throw std::invalid_argument("Capacity must be positive");
        }
    }

    bool empty() const { return size_ == 0; }
    bool full() const { return size_ == data_.size(); }
    size_t size() const { return size_; }
    size_t capacity() const { return data_.size(); }

    void push_front(const T& value) {
        if (full()) {
            throw std::overflow_error("Deque overflow");
        }
        front_ = (front_ - 1 + data_.size()) % data_.size();
        data_[front_] = value;
        ++size_;
    }

    void push_back(const T& value) {
        if (full()) {
            throw std::overflow_error("Deque overflow");
        }
        size_t idx = (front_ + size_) % data_.size();
        data_[idx] = value;
        ++size_;
    }

    void pop_front() {
        if (empty()) {
            throw std::underflow_error("Deque underflow");
        }
        front_ = (front_ + 1) % data_.size();
        --size_;
    }

    void pop_back() {
        if (empty()) {
            throw std::underflow_error("Deque underflow");
        }
        --size_;
    }

    const T& front() const {
        if (empty()) {
            throw std::underflow_error("Deque underflow");
        }
        return data_[front_];
    }

    const T& back() const {
        if (empty()) {
            throw std::underflow_error("Deque underflow");
        }
        size_t idx = (front_ + size_ - 1) % data_.size();
        return data_[idx];
    }

private:
    std::vector<T> data_;
    size_t front_;
    size_t size_;
};

/**
 * @brief Doubly linked-list deque.
 * Worst-case O(1) push and pop at both ends without resizing.
 */
template <typename T>
class LinkedDeque {
private:
    struct Node {
        T value;
        Node* prev;
        Node* next;
        Node(const T& val) : value(val), prev(nullptr), next(nullptr) {}
    };

public:
    LinkedDeque() : head_(nullptr), tail_(nullptr), size_(0) {}

    ~LinkedDeque() {
        clear();
    }

    LinkedDeque(const LinkedDeque&) = delete;
    LinkedDeque& operator=(const LinkedDeque&) = delete;

    void clear() {
        while (head_) {
            Node* old = head_;
            head_ = head_->next;
            delete old;
        }
        tail_ = nullptr;
        size_ = 0;
    }

    bool empty() const { return size_ == 0; }
    size_t size() const { return size_; }

    void push_front(const T& value) {
        Node* node = new Node(value);
        node->next = head_;
        if (head_) {
            head_->prev = node;
        } else {
            tail_ = node;
        }
        head_ = node;
        ++size_;
    }

    void push_back(const T& value) {
        Node* node = new Node(value);
        node->prev = tail_;
        if (tail_) {
            tail_->next = node;
        } else {
            head_ = node;
        }
        tail_ = node;
        ++size_;
    }

    void pop_front() {
        if (!head_) {
            throw std::underflow_error("Deque underflow");
        }
        Node* old = head_;
        head_ = head_->next;
        if (head_) {
            head_->prev = nullptr;
        } else {
            tail_ = nullptr;
        }
        delete old;
        --size_;
    }

    void pop_back() {
        if (!tail_) {
            throw std::underflow_error("Deque underflow");
        }
        Node* old = tail_;
        tail_ = tail_->prev;
        if (tail_) {
            tail_->next = nullptr;
        } else {
            head_ = nullptr;
        }
        delete old;
        --size_;
    }

    const T& front() const {
        if (!head_) {
            throw std::underflow_error("Deque underflow");
        }
        return head_->value;
    }

    const T& back() const {
        if (!tail_) {
            throw std::underflow_error("Deque underflow");
        }
        return tail_->value;
    }

private:
    Node* head_;
    Node* tail_;
    size_t size_;
};

/**
 * @brief Computes maximum value in every sliding window of size k in O(n) time.
 * Uses a monotonic deque to maintain decreasing candidates.
 */
inline std::vector<int> sliding_window_max(const std::vector<int>& nums, int k) {
    std::deque<int> dq; // stores indices
    std::vector<int> result;
    int n = static_cast<int>(nums.size());
    if (n == 0 || k <= 0 || k > n) return result;

    for (int i = 0; i < n; ++i) {
        // Remove indices outside current window
        while (!dq.empty() && dq.front() <= i - k) {
            dq.pop_front();
        }

        // Maintain decreasing order: discard candidates smaller than current
        while (!dq.empty() && nums[dq.back()] <= nums[i]) {
            dq.pop_back();
        }

        dq.push_back(i);

        // Record maximum once window reaches size k
        if (i >= k - 1) {
            result.push_back(nums[dq.front()]);
        }
    }

    return result;
}

} // namespace dsa

// ============================================================================
// Unit Tests
// ============================================================================

void test_circular_deque() {
    dsa::CircularDeque<int> dq(4);
    assert(dq.empty());

    dq.push_back(10);
    dq.push_back(20);
    dq.push_front(5);
    dq.push_front(1);
    assert(dq.full());
    assert(dq.size() == 4);
    assert(dq.front() == 1);
    assert(dq.back() == 20);

    dq.pop_front();
    assert(dq.front() == 5);
    dq.pop_back();
    assert(dq.back() == 10);

    dq.push_back(30);
    assert(dq.back() == 30);
    assert(dq.front() == 5);

    dq.pop_front();
    dq.pop_front();
    dq.pop_front();
    assert(dq.empty());
}

void test_linked_deque() {
    dsa::LinkedDeque<std::string> dq;
    dq.push_front("mid");
    dq.push_front("start");
    dq.push_back("end");
    assert(dq.size() == 3);
    assert(dq.front() == "start");
    assert(dq.back() == "end");

    dq.pop_front();
    assert(dq.front() == "mid");
    dq.pop_back();
    assert(dq.back() == "mid");
    dq.pop_front();
    assert(dq.empty());
}

void test_sliding_window_max() {
    std::vector<int> nums = {1, 3, -1, -3, 5, 3, 6, 7};
    int k = 3;
    auto res = dsa::sliding_window_max(nums, k);
    std::vector<int> expected = {3, 3, 5, 5, 6, 7};
    assert(res == expected);
}

int main() {
    std::cout << "Running Deque C++17 unit tests...\n";
    test_circular_deque();
    test_linked_deque();
    test_sliding_window_max();
    std::cout << "All Deque C++17 unit tests passed successfully!\n";
    return 0;
}
